packages/typescript-estree/src/node-utils.ts

```diff
@@ -649,16 +649,27 @@ export function convertTokens(ast: ts.SourceFile): TSESTree.Token[] {
   return result;
 }
 
-export interface TSError {
-  index: number;
-  lineNumber: number;
-  column: number;
-  message: string;
+export class TSError extends Error {
+  constructor(
+    message: string,
+    public fileName: string,
+    public index: number,
+    public lineNumber: number,
+    public column: number,
+  ) {
+    super(message);
+    Object.setPrototypeOf(this, TSError.prototype);
+    Object.defineProperty(this, 'name', {
+      value: new.target.name,
+      enumerable: false,
+      configurable: true,
+    });
```

Comment:
We target ES2017

https://github.com/typescript-eslint/typescript-eslint/blob/dbe5885c3102d2ce6dddfb37fab681115659bd31/tsconfig.base.json#L18

Which means `super(message)` is not changed and is left as a `super` call.

are you sure we still need this?

Reply:
yes we do, otherwise name will be Error instead of TSError

Reply:
can we just make it a readonly property?

What about the `setPrototypeOf` - do we need  that function call at all (that's what I was referring to mostly)

Reply:
we kinda do, and its even recommended in ts docs: https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work

Reply:
> we kinda do, and its even recommended in ts docs

But that's in the context of changes in TS2.1's transpilation for ES5 (see the section above it):
https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#generated-constructor-code-substitutes-the-return-value-of-super-calls-as-this

As mentioned - we target es2017, so there is no such transpilation:

<details>
<summary>es2017 target</summary>

```js
class TSError extends Error {
    constructor(message, fileName, index, lineNumber, column) {
        super(message); ////////// this is a pure `super` call //////////
        this.fileName = fileName;
        this.index = index;
        this.lineNumber = lineNumber;
        this.column = column;
        Object.setPrototypeOf(this, TSError.prototype);
        Object.defineProperty(this, 'name', {
            value: new.target.name,
            enumerable: false,
            configurable: true,
        });
    }
}
```

</details>

<details>
<summary>es5 target</summary>

```js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var TSError = /** @class */ (function (_super) {
    __extends(TSError, _super);
    function TSError(message, fileName, index, lineNumber, column) {
        var _newTarget = this.constructor;
        var _this = _super.call(this, message) || this; ////////// this is not a pure super call //////////
        _this.fileName = fileName;
        _this.index = index;
        _this.lineNumber = lineNumber;
        _this.column = column;
        Object.setPrototypeOf(_this, TSError.prototype);
        Object.defineProperty(_this, 'name', {
            value: _newTarget.name,
            enumerable: false,
            configurable: true,
        });
        return _this;
    }
    return TSError;
}(Error));
```

</details>

Reply:
```
Object.setPrototypeOf(this, TSError.prototype);
```
removed